snuba/state/rate_limit.py

```diff
@@ -0,0 +1,183 @@
+    +-----------------------------+--------------------------------+
+    | historical query window     | currently executing queries    |
+    +-----------------------------+--------------------------------+
+                                  ^
+                                 now
+    """
+
+    def __init__(self, request):
+        self._request = request
+        self.__did_run = False
+        self.__query_id = None
+        self.__bucket = None
+
+    @abstractmethod
+    def _get_rate_limit_params(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_rate_limit_name(self):
+        raise NotImplementedError
+
+    def __enter__(self):
+        rate_limit_params = self._get_rate_limit_params()
+
+        self.__bucket = '{}{}'.format(state.ratelimit_prefix, rate_limit_params.bucket)
+        self.__query_id = uuid.uuid4()
+        now = time.time()
+        bypass_rate_limit, rate_history_s = state.get_configs([
+            ('bypass_rate_limit', 0),
+            ('rate_history_sec', 3600)
+        ])
+
+        if bypass_rate_limit == 1:
+            return (True, 0, 0)
+
+        pipe = state.rds.pipeline(transaction=False)
+        pipe.zremrangebyscore(self.__bucket, '-inf', '({:f}'.format(now - rate_history_s))  # cleanup
+        pipe.zadd(self.__bucket, now + state.max_query_duration_s, self.__query_id)  # add query
+        if rate_limit_params.per_second_limit is None:
+            pipe.exists("nosuchkey")  # no-op if we don't need per-second
+        else:
+            pipe.zcount(self.__bucket, now - state.rate_lookback_s, now)  # get historical
+        if rate_limit_params.concurrent_limit is None:
+            pipe.exists("nosuchkey")  # no-op if we don't need concurrent
+        else:
+            pipe.zcount(self.__bucket, '({:f}'.format(now), '+inf')  # get concurrent
+
+        try:
+            _, _, historical, concurrent = pipe.execute()
+            historical = int(historical)
+            concurrent = int(concurrent)
+            self.__did_run = True
+        except Exception as ex:
+            logger.exception(ex)
+            return (True, 0, 0)  # fail open if redis is having issues
+
+        per_second = historical / float(state.rate_lookback_s)
+
+        stats = {
+            '{}_rate'.format(self._get_rate_limit_name()): per_second,
+            '{}_concurrent'.format(self._get_rate_limit_name()): concurrent,
+
+        }
+
+        Reason = namedtuple('reason', 'scope name val limit')
+        reasons = [
+            Reason(self._get_rate_limit_name(), 'concurrent', concurrent, rate_limit_params.concurrent_limit),
+            Reason(self._get_rate_limit_name(), 'per-second', per_second, rate_limit_params.per_second_limit),
+        ]
+
+        reason = next((r for r in reasons if r.limit is not None and r.val > r.limit), None)
+        error = reason and '{r.scope} {r.name} of {r.val:.0f} exceeds limit of {r.limit:.0f}'.format(r=reason)
+
+        return error, stats
```

Comment:
shouldn't we throw if we exceed the limit? In a context manager, if you cannot acquire the resource an exception is thrown generally, you do not enter the with block with an invalid context.